### PR TITLE
prepare j1587driver for use as a transport reassembler component

### PR DESCRIPTION
**Pull request by BenGardiner**

### CODE REVIEW

File: hv_networks/J1587Driver.py [hunk 1]
```diff
@@ -312,24 +321,30 @@ def make(self):
...
-    j1708_factory_singleton = factory
+    with factory_lock:
+        j1708_factory_singleton = factory
 
 
 def get_j1708_driver_factory():
-    return j1708_factory_singleton
+    with factory_lock:
+        a = j1708_factory_singleton
+    return a
```
Comment (salnoobie) [inline hv_networks/J1587Driver.py:337]: Is there a case `a` is not set here?

File: test_j1587_driver.py [hunk 2]
```diff
@@ -1,66 +1,96 @@
...
 
     def make(self):
-        if self.memo_fake_driver is None:
-            self.memo_fake_driver = FakeJ1708Driver()
-        return self.memo_fake_driver
+        with self.a_lock:
+            if self.memo_fake_driver is None:
+                self.memo_fake_driver = FakeJ1708Driver()
+            a = self.memo_fake_driver
+        return a
```
Comment (salnoobie) [inline test_j1587_driver.py:66]: Same thing. Unsure of behavior if `with` doesn't work out.

**Review by salnoobie — COMMENTED** [review]
Looks good. Couple of places have a potential weird condition if the returned value was not set after a `with`. I added comments to the lines as I'm unsure what will happen.

File: hv_networks/J1587Driver.py (hunk 1 above)
Comment (haystack-ia) [reply hv_networks/J1587Driver.py:337]: I think the only way that this function doesn't return is if it deadlocks waiting for `factory_lock`, right? Is there something I don't see?

As long as you use the lock responsibly (which this code appears to) it should be fine.

Comment (BenGardiner) [reply hv_networks/J1587Driver.py:337]: Yep. 
Use of the lock is here for publication guarantees of this assignment to the other threads reading the global (also under the lock). 

@salnoobie: `a` is alway (eventually) set.

File: test_j1587_driver.py (hunk 2 above)
Comment (BenGardiner) [reply test_j1587_driver.py:66]: there's no deadlocks (that I'm aware of, *gulp*) so `a` will always be (eventually) assigned.

**Comment by BenGardiner** [comment]
Thanks @salnoobie and @haystack-ia !

File: hv_networks/J1587Driver.py (hunk 1 above)
Comment (salnoobie) [reply hv_networks/J1587Driver.py:337]: Oh excellent!

**Comment by BenGardiner** [comment]
made a couple small changes after that last review and bump the version to 0.2 here. Planned to be use to add transport reassembly to pretty_j1587 here: https://github.com/ainfosec/pretty_j1587/pull/5

**Pull request merged**